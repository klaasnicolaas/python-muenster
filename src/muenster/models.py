"""Models for Open Data Platform of MÃ¼nster."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import pyproj


@dataclass
class Garage:
    """Object representing a garage."""

    name: str
    status: str
    parking_type: str
    free_space: int
    total_capacity: int
    availability_pct: float
    url: str

    longitude: float
    latitude: float

    @classmethod
    def from_dict(cls: type[Garage], data: dict[str, Any]) -> Garage:
        """Return a Garage object from a dictionary.

        Args:
        ----
            data: The data from the API.

        Returns:
        -------
            A Garage object.
        """
        attr = data["properties"]
        geo = data["geometry"]["coordinates"]

        # Convert the coordinates from Guass-Kruger (zone 3) to WGS84.
        gauss_kruger_wgs = pyproj.Transformer.from_crs(31467, 4326)
        converted = gauss_kruger_wgs.transform(geo[1], geo[0])

        return cls(
            name=attr.get("NAME"),
            status=attr.get("status"),
            parking_type=attr.get("type"),
            free_space=attr.get("parkingFree"),
            total_capacity=attr.get("parkingTotal"),
            availability_pct=round(
                (float(attr.get("parkingFree")) / float(attr.get("parkingTotal")))
                * 100,
                1,
            ),
            url=attr.get("URL"),
            longitude=converted[1],
            latitude=converted[0],
        )
