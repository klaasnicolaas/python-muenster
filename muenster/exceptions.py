"""Asynchronous Python client providing Open Data information of Muenster."""


class ODPMuensterError(Exception):
    """Generic Open Data Platform Muenster exception."""


class ODPMuensterConnectionError(ODPMuensterError):
    """Open Data Platform Muenster - connection error."""
