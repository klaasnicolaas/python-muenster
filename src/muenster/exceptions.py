"""Asynchronous Python client providing Open Data information of Münster."""


class ODPMuensterError(Exception):
    """Generic Open Data Platform Münster exception."""


class ODPMuensterConnectionError(ODPMuensterError):
    """Open Data Platform Münster - connection error."""
