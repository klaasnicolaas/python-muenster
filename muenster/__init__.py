"""Asynchronous Python client providing Open Data information of MÃ¼nster."""

from .exceptions import ODPMuensterConnectionError, ODPMuensterError
from .models import Garage
from .muenster import ODPMuenster, StadtMuenster

__all__ = [
    "ODPMuenster",
    "ODPMuensterConnectionError",
    "ODPMuensterError",
    "Garage",
    "StadtMuenster",
]
