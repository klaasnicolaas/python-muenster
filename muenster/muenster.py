"""Asynchronous Python client providing Open Data information of Münster."""
from __future__ import annotations

import asyncio
import json
import socket
from dataclasses import dataclass
from importlib import metadata
from typing import Any, cast

import async_timeout
from aiohttp import ClientError, ClientSession
from aiohttp.hdrs import METH_GET
from yarl import URL

from .exceptions import ODPMuensterConnectionError, ODPMuensterError
from .models import Garage


@dataclass
class ODPMuenster:
    """Main class for handling data fetchting from Open Data Platform of Münster."""

    request_timeout: float = 10.0
    session: ClientSession | None = None

    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Handle a request to the Open Data Platform API of Münster.

        Args:
        ----
            uri: Request URI, without '/', for example, 'status'
            method: HTTP method to use, for example, 'GET'
            params: Extra options to improve or limit the response.

        Returns:
        -------
            A Python dictionary (text) with the response from
            the Open Data Platform API of Münster.

        Raises:
        ------
            ODPMuensterConnectionError: Timeout occurred while
                connecting to the Open Data Platform API.
            ODPMuensterError: If the data is not valid.
        """
        version = metadata.version(__package__)
        url = URL.build(
            scheme="https",
            host="opendata.stadt-muenster.de",
            path="/dataset/",
        ).join(URL(uri))

        headers = {
            "Accept": "application/json",
            "User-Agent": f"PythonODPMuenster/{version}",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    params=params,
                    headers=headers,
                    ssl=True,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to the Open Data Platform API."
            raise ODPMuensterConnectionError(
                msg,
            ) from exception
        except (ClientError, socket.gaierror) as exception:
            msg = "Error occurred while communicating with Open Data Platform API."
            raise ODPMuensterConnectionError(
                msg,
            ) from exception

        content_type = response.headers.get("Content-Type", "")
        if "application/json" not in content_type:
            text = await response.text()
            msg = "Unexpected content type response from the Open Data Platform API"
            raise ODPMuensterError(
                msg,
                {"Content-Type": content_type, "Response": text},
            )

        return cast(dict[str, Any], await response.json())

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> ODPMuenster:
        """Async enter.

        Returns
        -------
            The Open Data Platform Münster object.
        """
        return self

    async def __aexit__(self, *_exc_info: Any) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.
        """
        await self.close()


@dataclass
class StadtMuenster:
    """Main class for handling data fetchting from Stadt of Münster."""

    request_timeout: float = 10.0
    session: ClientSession | None = None

    _close_session: bool = False

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        params: dict[str, Any] | None = None,
    ) -> Any:
        """Handle a request to the Open Data Platform API of Münster.

        Args:
        ----
            uri: Request URI, without '/', for example, 'status'
            method: HTTP method to use, for example, 'GET'
            params: Extra options to improve or limit the response.

        Returns:
        -------
            A Python dictionary (text) with the response from
            the Open Data Platform API of Münster.

        Raises:
        ------
            ODPMuensterConnectionError: Timeout occurred while
                connecting to the Open Data Platform API.
            ODPMuensterError: If the data is not valid.
        """
        version = metadata.version(__package__)
        url = URL.build(scheme="https", host="stadt-muenster.de", path="/").join(
            URL(uri),
        )

        headers = {
            "Accept": "text/html",
            "User-Agent": f"PythonStadtMuenster/{version}",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    params=params,
                    headers=headers,
                    ssl=True,
                )
                response.raise_for_status()
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to the Open Data Platform API."
            raise ODPMuensterConnectionError(
                msg,
            ) from exception
        except (ClientError, socket.gaierror) as exception:
            msg = "Error occurred while communicating with Open Data Platform API."
            raise ODPMuensterConnectionError(
                msg,
            ) from exception

        content_type = response.headers.get("Content-Type", "")
        if "text/html" not in content_type:
            text = await response.text()
            msg = "Unexpected content type response from the Open Data Platform API"
            raise ODPMuensterError(
                msg,
                {"Content-Type": content_type, "Response": text},
            )

        return json.loads(await response.text())

    async def garages(self) -> list[Garage]:
        """Get list of parking garages.

        Returns
        -------
            A list of Garage objects.
        """
        results: list[Garage] = []
        locations = await self._request(
            "index.php",
            params={"id": "10910"},
        )

        for item in locations["features"]:
            results.append(Garage.from_dict(item))
        return results

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> StadtMuenster:
        """Async enter.

        Returns
        -------
            The Stadt Münster object.
        """
        return self

    async def __aexit__(self, *_exc_info: Any) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.
        """
        await self.close()
