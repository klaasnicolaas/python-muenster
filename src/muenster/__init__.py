"""Asynchronous Python client providing Open Data information of MÃ¼nster."""

from .exceptions import ODPMuensterConnectionError, ODPMuensterError
from .models import Garage
from .muenster import ODPMuenster, StadtMuenster

__all__ = [
    "Garage",
    "ODPMuenster",
    "ODPMuensterConnectionError",
    "ODPMuensterError",
    "StadtMuenster",
]
