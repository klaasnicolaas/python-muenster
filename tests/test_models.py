"""Test the models."""

from __future__ import annotations

from aresponses import ResponsesMockServer

from muenster import Garage, StadtMuenster

from . import load_fixtures


async def test_all_garages(
    aresponses: ResponsesMockServer, stadt_muenster_client: StadtMuenster
) -> None:
    """Test all garages function."""
    aresponses.add(
        "stadt-muenster.de",
        "/index.php",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "text/html"},
            text=load_fixtures("garages.json"),
        ),
    )
    spaces: list[Garage] = await stadt_muenster_client.garages()
    assert spaces is not None
    for item in spaces:
        assert isinstance(item, Garage)
        assert item.url is not None
        assert item.longitude is not None
        assert item.latitude is not None
        assert isinstance(item.longitude, float)
        assert isinstance(item.latitude, float)
