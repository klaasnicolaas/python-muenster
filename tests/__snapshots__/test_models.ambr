# serializer version: 1
# name: test_all_garages
  list([
    Garage(name='Parkhaus Theater', status='frei', parking_type='Parkhaus', free_space=537, total_capacity='748', availability_pct=71.8, url='https://www.stadt-muenster.de/tiefbauamt/parkleitsystem/parkhaeuser/detailansicht/parkhaus/1.html', longitude=7.626790040283135, latitude=51.96599313816084),
    Garage(name='Parkplatz Hörster Platz', status='frei', parking_type='Parkplatz', free_space=91, total_capacity='202', availability_pct=45.0, url='https://www.stadt-muenster.de/tiefbauamt/parkleitsystem/parkhaeuser/detailansicht/parkhaus/2.html', longitude=7.63168867787138, latitude=51.964243549926614),
    Garage(name='Parkhaus Alter Steinweg', status='frei', parking_type='Parkhaus', free_space=166, total_capacity='390', availability_pct=42.6, url='https://www.stadt-muenster.de/tiefbauamt/parkleitsystem/parkhaeuser/detailansicht/parkhaus/3.html', longitude=7.632386884741256, latitude=51.96233720813437),
    Garage(name='Parkplatz Busparkplatz', status='frei', parking_type='Parkplatz', free_space=53, total_capacity='63', availability_pct=84.1, url='https://www.stadt-muenster.de/tiefbauamt/parkleitsystem/parkhaeuser/detailansicht/parkhaus/4.html', longitude=7.61680229612901, latitude=51.96571449455512),
    Garage(name='Parkplatz Schlossplatz Nord', status='frei', parking_type='Parkplatz', free_space=450, total_capacity='450', availability_pct=100.0, url='https://www.stadt-muenster.de/tiefbauamt/parkleitsystem/parkhaeuser/detailansicht/parkhaus/5.html', longitude=7.616959853461621, latitude=51.96437712646424),
    Garage(name='Parkplatz Schlossplatz Süd', status='frei', parking_type='Parkplatz', free_space=166, total_capacity='460', availability_pct=36.1, url='https://www.stadt-muenster.de/tiefbauamt/parkleitsystem/parkhaeuser/detailansicht/parkhaus/6.html', longitude=7.6168499026422, latitude=51.96274899957397),
    Garage(name='Parkhaus Aegidii', status='frei', parking_type='Parkhaus', free_space=490, total_capacity='750', availability_pct=65.3, url='https://www.stadt-muenster.de/tiefbauamt/parkleitsystem/parkhaeuser/detailansicht/parkhaus/7.html', longitude=7.6230306668972245, latitude=51.96046654876251),
    Garage(name='Parkplatz Georgskommende', status='frei', parking_type='Parkhaus', free_space=190, total_capacity='272', availability_pct=69.9, url='https://www.stadt-muenster.de/tiefbauamt/parkleitsystem/parkhaeuser/detailansicht/parkhaus/8.html', longitude=7.619172622661307, latitude=51.959603471601845),
    Garage(name='Parkhaus Münster-Arkaden', status='geschlossen', parking_type='Parkhaus', free_space=0, total_capacity='150', availability_pct=0.0, url='https://www.stadt-muenster.de/tiefbauamt/parkleitsystem/parkhaeuser/detailansicht/parkhaus/9.html', longitude=7.626485216249938, latitude=51.959778851561595),
    Garage(name='Parkhaus Karstadt', status='frei', parking_type='Parkhaus', free_space=110, total_capacity='183', availability_pct=60.1, url='https://www.stadt-muenster.de/tiefbauamt/parkleitsystem/parkhaeuser/detailansicht/parkhaus/10.html', longitude=7.6304115214833566, latitude=51.96127164447771),
    Garage(name='Parkhaus Stubengasse', status='frei', parking_type='Parkhaus', free_space=140, total_capacity='318', availability_pct=44.0, url='https://www.stadt-muenster.de/tiefbauamt/parkleitsystem/parkhaeuser/detailansicht/parkhaus/11.html', longitude=7.629794561445889, latitude=51.96003310823958),
    Garage(name='Parkhaus Bremer Platz', status='frei', parking_type='Parkhaus', free_space=149, total_capacity='354', availability_pct=42.1, url='https://www.stadt-muenster.de/tiefbauamt/parkleitsystem/parkhaeuser/detailansicht/parkhaus/12.html', longitude=7.637656107075433, latitude=51.95794020639265),
    Garage(name='Parkhaus Engelenschanze', status='frei', parking_type='Parkhaus', free_space=276, total_capacity='480', availability_pct=57.5, url='https://www.stadt-muenster.de/tiefbauamt/parkleitsystem/parkhaeuser/detailansicht/parkhaus/13.html', longitude=7.631394483247177, latitude=51.95574643864072),
    Garage(name='Parkhaus Bahnhofstraße', status='frei', parking_type='Parkhaus', free_space=181, total_capacity='311', availability_pct=58.2, url='https://www.stadt-muenster.de/tiefbauamt/parkleitsystem/parkhaeuser/detailansicht/parkhaus/14.html', longitude=7.632952075623044, latitude=51.955234232717444),
    Garage(name='Parkhaus Cineplex', status='frei', parking_type='Parkhaus', free_space=487, total_capacity='590', availability_pct=82.5, url='https://www.stadt-muenster.de/tiefbauamt/parkleitsystem/parkhaeuser/detailansicht/parkhaus/15.html', longitude=7.636025382313851, latitude=51.94983209759471),
    Garage(name='Parkhaus PH Stadthaus 3', status='frei', parking_type='Parkhaus', free_space=241, total_capacity='372', availability_pct=64.8, url='https://www.stadt-muenster.de/tiefbauamt/parkleitsystem/parkhaeuser/detailansicht/parkhaus/16.html', longitude=7.640941227445868, latitude=51.94936761869015),
  ])
# ---
